### PR TITLE
Change console keyboard layout

### PR DESCRIPTION
**Pull request by pmos69**
Effective after reboot.

### CODE REVIEW

**Comment by XECDesign** [comment]
It appears to work as is. What problem does this fix?

**Comment by pmos69** [comment]
Changes keyboard in console, not just in X.
Although the console change only takes effect after a reboot.

**Comment by XECDesign** [comment]
The current raspi-config does it for console as well.

**Comment by pmos69** [comment]
Test it.
Use raspi-config to change to a different keyboard layout.
Then do ctrl-alt-f1 to change to console.
Check that the keyboard layout was not changed.
Reboot and enter console again. Check the layout hasn't changed..
In console, execute "setupcon". Check the console keyboard layout was changed.
The proposed change in raspi-config uses the --save-only option because setupcon only works imediately when executed from the console, not from within X.

**Comment by pmos69** [comment]
(When executing setupcon from console you have to use sudo)

**Comment by XECDesign** [comment]
Tested it before posting using the latest upgraded image.

**Comment by pmos69** [comment]
Weird.
Made the change last night after bumping into the issue. Alto with an upgraded image.
I'll check again tonight after work. No point in fixing a non existant issue.

**Comment by pmos69** [comment]
Hi,

Just tested again:
Configuring keyboard in raspi-config only changed the keyboard layout in X terminals and applications. All console tty terminals remained with the old layout, even after a reboot.
How can you get different results?
To clear any doubts, I installed a fresh system on a new SD, then updated the system:

> pi@raspberrypi:~ $ sudo apt-get update && sudo apt-get dist-upgrade
> Hit:1 http://mirrordirector.raspbian.org/raspbian stretch InRelease
> Hit:2 http://archive.raspberrypi.org/debian stretch InRelease
> Reading package lists... Done
> Reading package lists... Done
> Building dependency tree       
> Reading state information... Done
> Calculating upgrade... Done
> 0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
> pi@raspberrypi:~ $ uname -a
> Linux raspberrypi 4.9.41-v7+ #1023 SMP Tue Aug 8 16:00:15 BST 2017 armv7l GNU/Linux
> pi@raspberrypi:~ $ cat /etc/debian_version
> 9.1

This is a Raspberry Pi 3.
Are you sure you don't have any modification on other system scripts to enable updating the console keyboard layout at boot, for example?

**Comment by pmos69** [comment]
Crap: A second reboot and the console keyboard layout changed.
Strange behaviour.

**Comment by pmos69** [comment]
The 2 reboots for the change to take effect in console seems to be by design, from what I've seen of /etc/init.d/console-setup.sh and /lib/console-setup/console-setup.sh

The revised commit makes the change effective immediately. No reboots required.

**Comment by XECDesign** [comment]
Having the change take immediate effect is definitely a great improvement.
Given that raspi-config must run as root, I think you can simplify that command a bit.

In my testing, just running `setupcon -k` is all that's needed and works over ssh too (which was an issue under jessie).

**Comment by XECDesign** [comment]
Actually, when I was testing, I was using `--force` as well. Not sure if it's required for setupcon to work over ssh or not.

**Comment by pmos69** [comment]
Of course. Your're right. Nice.
Fix commited.

File: raspi-config [hunk 1]
```diff
@@ -293,6 +293,7 @@ do_configure_keyboard() {
   dpkg-reconfigure keyboard-configuration &&
   printf "Reloading keymap. This may take a short while\n" &&
   invoke-rc.d keyboard-setup start || return $?
+  setsid sh -c 'exec setupcon -k --force <> /dev/tty1 >&0 2>&1'
```
Comment (XECDesign) [inline raspi-config]: Why can't this whole line just be `setupcon -k --force`?

File: raspi-config [hunk 2]
```diff
@@ -1454,6 +1455,7 @@ do_internationalisation_menu() {
   FUN=$(whiptail --title "Raspberry Pi Software Configuration Tool (raspi-config)" --menu "Localisation Options" $WT_HEIGHT $WT_WIDTH $WT_MENU_HEIGHT --cancel-button Back --ok-button Select \
     "I1 Change Locale" "Set up language and regional settings to match your location" \
     "I2 Change Timezone" "Set up timezone to match your location" \
+    "I3 Change Keyboard Layout" "Set the keyboard layout to match your keyboard" \
```
Comment (XECDesign) [inline raspi-config]: You can take out the whole `if` block and is_ssh command, right?
Sorry to be nit-picky.

File: raspi-config (hunk 1 above)
Comment (pmos69) [reply raspi-config]: I tried it, but I didn't get consistent results.Sometimes the console layout would change immediatelly, sometimes it didn't, especially when doing several changes one after another.

Sending the setupcon command to be executed in tty1 seems to work immediately every time, as that's the expected behaviour of setupcon when executed at the console.

I know back-to-back changes in keyboard layout are probably not a very common scenario, but even so I would think a consistent behaviour is always better, unless there's a reason not to do it this way, of course.

File: raspi-config (hunk 2 above)
Comment (pmos69) [reply raspi-config]: Yup. Just pushed the code I was using for testing and didn't clean it. Lazy.
fixed

**Comment by XECDesign** [comment]
Alright, thanks for the fix. Would be nice to know why you weren't getting consistent results and maybe I'll take a closer look later, but this is good for now.

**Pull request merged**